lib/doorkeeper/oauth/error_response_type_request.rb

```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+module Doorkeeper
+  module OAuth
+    class ErrorResponseTypeRequest
+      attr_accessor :pre_auth
+
+      def initialize(pre_auth)
+        @pre_auth = pre_auth
+      end
+
+      def authorize
+        @response = if pre_auth.error == :invalid_request
```

Comment:
I see a lot of similar checks in the PR. My opinion is not to ask, but tell (*tell, don't ask*).

So it'd be better to have something like

```ruby
pre_auth.build_error_response
```

And hide this logic under this method or to introduce a `ErrorResponseBuilder` that will encapsulate the logic of building error response in it (this way I like more btw)

Reply:
Thank for your suggestion. Seem I can decrease the some complicate parts of this pull  
 request.

Reply:
@nbulaj 
Indeed `pre_auth` itself has `error_response` method to return ErrorResponse. So I fix and use it to return error_response for above case. 
Besides, I see that current logic of `authorize_response` method at authorizations_controller.rb is a bit confusing, because it continues processing even `pre_auth` is not authorizable.
https://github.com/doorkeeper-gem/doorkeeper/blob/6eda02d7b0e9971a71850aff49737e1cf141d2be/app/controllers/doorkeeper/authorizations_controller.rb#L87-L95
So I just return error_response immediately if pre_auth is not authorizable, so it make sure that the followed processing is always authorize. (not have error)
```ruby
def authorize_response
  @authorize_response ||= begin
    authorizable = pre_auth.authorizable?
    return pre_auth.error_response unless authorizable

    before_successful_authorization
    auth = strategy.authorize
    after_successful_authorization
    auth
  end
end
```
All the fixes are in this commit: https://github.com/doorkeeper-gem/doorkeeper/pull/1277/commits/9eeb70c9ba6b1f12e1caae6863b2eeaa64529205

Reply:
I like it :)